django/http/response.py

```diff
@@ -406,28 +408,41 @@ def _set_streaming_content(self, value):
         if hasattr(filelike, 'close'):
             self._closable_objects.append(filelike)
         value = iter(lambda: filelike.read(self.block_size), b'')
-        self.set_headers(filelike)
+        self._set_headers(filelike)
         super()._set_streaming_content(value)
 
-    def set_headers(self, filelike):
+    def _set_headers(self, filelike):
         """
         Set some common response headers (Content-Length, Content-Type, and
         Content-Disposition) based on the `filelike` response content.
         """
-        encoding_map = {
-            'bzip2': 'application/x-bzip',
-            'gzip': 'application/gzip',
-            'xz': 'application/x-xz',
-        }
-        filename = getattr(filelike, 'name', None)
-        filename = filename if (isinstance(filename, str) and filename) else self.filename
-        if os.path.isabs(filename):
-            self['Content-Length'] = os.path.getsize(filelike.name)
-        elif hasattr(filelike, 'getbuffer'):
+        filename = getattr(filelike, 'name', '')
+        filename = filename if isinstance(filename, str) else ''
+        if hasattr(filelike, 'getbuffer'):
             self['Content-Length'] = filelike.getbuffer().nbytes
-
-        if self.get('Content-Type', '').startswith('text/html'):
+        elif hasattr(filelike, 'seek'):
+            try:
+                filelike.seek(0)
+            except io.UnsupportedOperation:
+                pass
+            else:
+                if hasattr(filelike, 'tell'):
+                    filelike.seek(0, 2)
+                    self['Content-Length'] = filelike.tell()
+                else:
+                    self['Content-Length'] = sum(iter(lambda: len(filelike.read(self.block_size)), 0))
+                filelike.seek(0)
+        if 'Content-Length' not in self and os.path.exists(filename):
+            self['Content-Length'] = os.path.getsize(filename)
+
+        filename = os.path.basename(self.filename or filename)
+        if self._default_content_type:
             if filename:
+                encoding_map = {
```

Comment:
If we're moving `encoding_map` it may as well go below the comment, where it's used. (And perhaps we don't need to temporary variable then?)

Reply:
i moved it to be at the beginning of the block in which it's being used, so it should be the same performance-wise, while still relatively keeping the _declaring constants_ - _logic_ structure.

yes, we indeed do not need that variable, although i choose to not write it this way because i personally find it to be pretty ugly... but ok.

Reply:
If you prefer it, keep it for now. We can decide later. 🙂

Reply:
If you prefer it, keep it for now. We can decide later. 🙂

Reply:
Too late. :P